File: asd-millennio-backend/src/schemas/tesseramento.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, EmailStr, field_validator


class SocioOnboarding(BaseModel):
    """Auto-creazione del profilo adulto al primo accesso (POST /soci/me)."""
    nome: str
    cognome: str
    data_nascita: date
    codice_fiscale: str
    indirizzo: str | None = None
    telefono: str | None = None
    # email non richiesta: si usa quella del token Keycloak

    @field_validator("codice_fiscale")
    @classmethod
    def cf_up(cls, v: str) -> str:
        return v.upper().strip()


class MinoreCreate(BaseModel):
    """Il tutore aggiunge un figlio minorenne (POST /soci/me/minori)."""
    nome: str
    cognome: str
    data_nascita: date
    codice_fiscale: str
    indirizzo: str | None = None

    @field_validator("codice_fiscale")
    @classmethod
    def cf_up(cls, v: str) -> str:
        return v.upper().strip()


class DocumentoResponse(BaseModel):
    id: UUID
    tipo: str
    filename: str
    content_type: str
    size_bytes: int
    created_at: datetime

    model_config = {"from_attributes": True}


class TesseramentoRequest(BaseModel):
    """Avvio del tesseramento per una categoria (POST /soci/me/tesseramento)."""
    categoria: str  # 'volley','badminton','kung_fu','pickleball','sostenitore'

    @field_validator("categoria")
    @classmethod
    def cat_valida(cls, v: str) -> str:
        validi = {"volley", "badminton", "kung_fu", "pickleball", "sostenitore"}
        if v not in validi:
            raise ValueError(f"Categoria non valida: {v}")
        return v


class TesseramentoCheckoutResponse(BaseModel):
    tessera_id: UUID
    pagamento_id: UUID
    stripe_checkout_url: str
    importo_eur: float


# ── verifica staff ────────────────────────────────────────────────────────────

class DocumentoBreve(BaseModel):
    id: UUID
    tipo: str
    filename: str


class SocioBreve(BaseModel):
    id: UUID
    nome: str
    cognome: str
    codice_fiscale: str
    is_minor: bool


class TesseramentoDaVerificare(BaseModel):
    tessera_id: UUID
    numero_tessera: str
    categoria: str
    anno_sportivo: str | None
    verifica_scadenza: datetime | None
    importo_eur: float | None
    socio: SocioBreve
    tutore: SocioBreve | None
    documenti: list[DocumentoBreve]


# ── quote (dirigenza) ─────────────────────────────────────────────────────────

class QuotaCreate(BaseModel):
    categoria: str
    is_minore: bool = False
    importo_eur: float
    anno_sportivo: str
    note: str | None = None

    @field_validator("categoria")
    @classmethod
    def cat_valida(cls, v: str) -> str:
        validi = {"volley", "badminton", "kung_fu", "pickleball", "sostenitore"}
        if v not in validi:
            raise ValueError(f"Categoria non valida: {v}")
        return v


class QuotaUpdate(BaseModel):
    importo_eur: float | None = None
    attivo: bool | None = None
    note: str | None = None


class QuotaResponse(BaseModel):
    id: UUID
    categoria: str
    is_minore: bool
    importo_eur: float
    anno_sportivo: str
    attivo: bool
    note: str | None

    model_config = {"from_attributes": True}

File: asd-millennio-frontend/lib/api/tesseramento.ts
import { apiClient } from "./client";
import type { Socio } from "./soci";

export interface OnboardingData {
  nome: string;
  cognome: string;
  data_nascita: string; // YYYY-MM-DD
  codice_fiscale: string;
  indirizzo?: string;
  telefono?: string;
}

export interface MinoreData {
  nome: string;
  cognome: string;
  data_nascita: string;
  codice_fiscale: string;
  indirizzo?: string;
}

export interface DocumentoResponse {
  id: string;
  tipo: string;
  filename: string;
  content_type: string;
  size_bytes: number;
  created_at: string;
}

export interface CheckoutResponse {
  tessera_id: string;
  pagamento_id: string;
  stripe_checkout_url: string;
  importo_eur: number;
}

export interface Quota {
  id: string;
  categoria: string;
  is_minore: boolean;
  importo_eur: number;
  anno_sportivo: string;
  attivo: boolean;
  note: string | null;
}

export interface TesseramentoDaVerificare {
  tessera_id: string;
  numero_tessera: string;
  categoria: string;
  anno_sportivo: string | null;
  verifica_scadenza: string | null;
  importo_eur: number | null;
  socio: { id: string; nome: string; cognome: string; codice_fiscale: string; is_minor: boolean };
  tutore: { id: string; nome: string; cognome: string } | null;
  documenti: { id: string; tipo: string; filename: string }[];
}

// ── socio (self) ────────────────────────────────────────────────────────────

export async function creaProfilo(data: OnboardingData): Promise<Socio> {
  const { data: res } = await apiClient.post<Socio>("/soci/me", data);
  return res;
}

export async function caricaDocumento(
  tipo: "identita" | "tutela",
  file: File,
  socioId?: string,
): Promise<DocumentoResponse> {
  const form = new FormData();
  form.append("file", file);
  const { data } = await apiClient.post<DocumentoResponse>("/soci/me/documenti", form, {
    params: { tipo, socio_id: socioId },
    headers: { "Content-Type": "multipart/form-data" },
  });
  return data;
}

export async function avviaTesseramento(categoria: string): Promise<CheckoutResponse> {
  const { data } = await apiClient.post<CheckoutResponse>("/soci/me/tesseramento", { categoria });
  return data;
}

// ── minori ──────────────────────────────────────────────────────────────────

export async function aggiungiMinore(data: MinoreData): Promise<Socio> {
  const { data: res } = await apiClient.post<Socio>("/soci/me/minori", data);
  return res;
}

export async function fetchMieiMinori(): Promise<Socio[]> {
  const { data } = await apiClient.get<Socio[]>("/soci/me/minori");
  return data;
}

export async function avviaTesseramentoMinore(minoreId: string, categoria: string): Promise<CheckoutResponse> {
  const { data } = await apiClient.post<CheckoutResponse>(
    `/soci/me/minori/${minoreId}/tesseramento`,
    { categoria },
  );
  return data;
}

// ── consensi (riuso endpoint esistente) ───────────────────────────────────────

export async function firmaConsensoSocio(
  socioId: string,
  tipo: "privacy" | "trattamento_dati" | "foto_video" | "marketing",
  firmatoDa?: string,
): Promise<void> {
  await apiClient.post(`/soci/${socioId}/consensi`, {
    socio_id: socioId,
    tipo,
    testo_versione: "v1",
    firmato_da: firmatoDa,
  });
}

// ── staff: verifiche ──────────────────────────────────────────────────────────

export async function fetchVerifiche(): Promise<TesseramentoDaVerificare[]> {
  const { data } = await apiClient.get<TesseramentoDaVerificare[]>("/soci/verifiche");
  return data;
}

export async function confermaVerifica(tesseraId: string): Promise<void> {
  await apiClient.post(`/soci/verifiche/${tesseraId}/conferma`);
}

export async function rifiutaVerifica(tesseraId: string): Promise<void> {
  await apiClient.post(`/soci/verifiche/${tesseraId}/rifiuta`);
}

export async function scaricaDocumento(docId: string): Promise<void> {
  const res = await apiClient.get(`/soci/documenti/${docId}`, { responseType: "blob" });
  const url = window.URL.createObjectURL(res.data as Blob);
  window.open(url, "_blank");
  setTimeout(() => window.URL.revokeObjectURL(url), 10000);
}

// ── dirigenza: quote ──────────────────────────────────────────────────────────

export async function fetchQuote(annoSportivo?: string): Promise<Quota[]> {
  const { data } = await apiClient.get<Quota[]>("/dirigenza/quote-tessera", {
    params: annoSportivo ? { anno_sportivo: annoSportivo } : undefined,
  });
  return data;
}

export async function creaQuota(payload: {
  categoria: string;
  is_minore: boolean;
  importo_eur: number;
  anno_sportivo: string;
  note?: string;
}): Promise<Quota> {
  const { data } = await apiClient.post<Quota>("/dirigenza/quote-tessera", payload);
  return data;
}

export async function aggiornaQuota(
  id: string,
  payload: { importo_eur?: number; attivo?: boolean; note?: string },
): Promise<Quota> {
  const { data } = await apiClient.put<Quota>(`/dirigenza/quote-tessera/${id}`, payload);
  return data;
}
